olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -240,31 +255,39 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
 }
 
 NetworkAsyncCallback GetRetryCallback(
-    int current_try, int current_backdown_period, const RetrySettings& settings,
-    const NetworkAsyncCallback& callback,
+    int current_try, std::chrono::milliseconds current_backdown_period,
+    std::chrono::milliseconds accumulated_wait_time,
+    const RetrySettings& settings, const NetworkAsyncCallback& callback,
     const std::shared_ptr<http::NetworkRequest>& network_request,
     std::weak_ptr<http::Network> network,
     const std::weak_ptr<CancellationContext>& weak_cancel_context) {
   ++current_try;
   return [=](HttpResponse response) {
+    const auto max_wait_time =
+        std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::seconds(settings.timeout));
     if (current_try > settings.max_attempts ||
-        !settings.retry_condition(response)) {
+        !settings.retry_condition(response) ||
+        accumulated_wait_time >= max_wait_time) {
       callback(std::move(response));
     } else {
       // TODO there must be a better way
-      std::this_thread::sleep_for(
-          std::chrono::milliseconds(current_backdown_period));
+      const auto actual_wait_time = std::min(
+          current_backdown_period, max_wait_time - accumulated_wait_time);
+      std::this_thread::sleep_for(actual_wait_time);
+
+      const auto next_backdown_period =
```

Comment:
```suggestion
      const auto next_wait_time =
```
The variable name contradicts the function and other variable names